core/src/main/java/org/infinispan/util/concurrent/BlockingHandler.java

```diff
@@ -0,0 +1,158 @@
+    * @param stage stage that may or may not be complete to apply
+    * @param function the blocking function
+    * @param traceId an identifier that can be used to tell in a trace when an operation moves between threads
+    * @param <I> input value type to the function
+    * @param <O> output value type after being transformed via function
+    * @return a stage that when complete will contain the value returned from the function or a throwable
+    */
+   <I, O> CompletionStage<O> thenApplyBlocking(CompletionStage<? extends I> stage,
+         Function<? super I, ? extends O> function, Object traceId);
+
+   /**
+    * Replacement for {@link CompletionStage#whenComplete(BiConsumer)} when the consumer is blocking to ensure the operation
+    * is performed upon a blocking thread and handles chained stages being invoked on non blocking thread as needed.
+    * @param stage stage that may or may not be complete to apply
+    * @param biConsumer the blocking biConsumer
+    * @param traceId an identifier that can be used to tell in a trace when an operation moves between threads
+    * @param <V> stage value type
+    * @return a stage that is complete when the biConsumer is but retains the results from the original stage
+    */
+   <V> CompletionStage<V> whenCompleteBlocking(CompletionStage<V> stage,
+         BiConsumer<? super V, ? super Throwable> biConsumer, Object traceId);
+
+   /**
+    * When the provided stage is complete, continue the completion chain of the returned CompletionStage on the
+    * supplied executor. If tracing is enabled a trace message is printed using the object as an identifier to more
+    * easily track the transition between threads.
+    * <p>
+    * This method is useful when an asynchronous computation completes and you do not want to run further processing
+    * on the thread that returned it. An example may be that some blocking operation is performed on a special blocking
+    * thread pool. However when the blocking operation completes we will want to continue the processing of that result
+    * in a thread pool that is for computational tasks.
+    * <p>
+    * If the supplied stage is already completed when invoking this command, this will return an already completed
+    * stage, which means any additional dependent stages will run in the invoking thread.
+    * @param <V> return value type of the supplied stage
+    * @param delay the stage to delay the continuation until complete
+    * @param traceId the id to print when tracing is enabled
+    * @return a CompletionStage that when depended upon will run any callback in the supplied executor
+    */
+   <V> CompletionStage<V> continueOnNonBlockingThread(CompletionStage<V> delay, Object traceId);
+
+   /**
+    * Provided a publisher that is known to block, this will ensure that the publisher is subscribed on the
+    * blocking thread and any values published will be on a non blocking thread. Note that if a blocking thread
```

Comment:
Feels like there's something missing after "any values published will be"

Reply:
observed :)